content/post/codesent.md

```diff
@@ -0,0 +1,381 @@
+
+* [src-d/go-git](https://github.com/src-d/go-git) to read Git repositories
+* [src-d/hercules](https://github.com/src-d/hercules) to analyze the commit tree
+* [Babelfish](https://doc.bblf.sh) to extract the comments
+* [vmarkovtsev/BiDiSentiment](https://github.com/vmarkovtsev/BiDiSentiment) to perform sentiment analysis (runs on Tensorflow)
+* Python to train the ML model and Go to apply it.
+
+I will describe each one in detail now. Let's start with go-git.
+
+{{% caption src="/post/codesent/go-git.png" %}}
+go-git logo by Ricardo Baeta.
+{{% /caption %}}
+
+We started it back in 2015 as an "internal open source" project to fulfill our tasks of source code
+retrieval and analysis. This is an example of the typical software project workflow at source{d}
+today: we go open source from day 0 and see what happens next. This time go-git became a huge success
+with more than 2400 stars and several companies already using it in production. Although it lacks
+some of the [features](https://github.com/src-d/go-git/blob/master/COMPATIBILITY.md) of "big brothers",
+it is performant and highly extensible. Finally, the maintainers are experienced code maniacs and
+go-git is a good example of idiomatic and clean Go.
+
+Hercules [appeared](https://blog.sourced.tech/post/hercules/) in December 2016 as an attempt
+to speed up Erik Bernhardsson's [Git-of-Theseus](https://erikbern.com/2016/12/05/the-half-life-of-code.html)
+by reimplementing it in Go and go-git. Indeed, it is an order of magnitude faster now.
+
+{{% caption src="/post/codesent/linux.png" %}}
+Line burndown of the Linux kernel - each band is one year and the thickness shows the ratio of lines
+which are not overwritten yet.
+{{% /caption %}}
+
+Then I added code ownership...
+
+{{% caption src="/post/codesent/emberjs_people.png" %}}
+Line ownership in Ember.js. Notice how one of devs appeared and conquered a half of the project.
+{{% /caption %}}
+
+...line overwrite matrix...
+
+{{% caption src="/post/codesent/wireshark_churn_matrix_black.png" %}}
+"Code wars" in Wireshark: XY cell's intensity is proportional to the number of times developer X's
+lines were overwritten with developer Y's.
+{{% /caption %}}
+
+...and even structural coupling.
+
+{{% caption src="/post/codesent/jinja_black.png" %}}
+Structural clusters in Jinja2. Each dot is a function and each pair is closer proportionally to
+the number of times those two functions appeared in the same commit.
+{{% /caption %}}
+
+Under the cover, Hercules contains the DAG resolution engine which frees users from manually
+designing the analysis pipeline. It also leverages `pkg/plugin` to allow external extensions.
+
+We pass over to comment extraction. The most straightforward way to extract comments is to
+take a syntax highlighter with a bunch of regular expression rules. We will get into trouble once
+we want to distinguish docstrings from regular comments, so we need to parse the code fairly.
```

Comment:
Not sure whether the word fairly is adequate in this context.
Maybe "correctly"?